Prevent dangling cat-file calls

zeripath opened the pull request:
If an `os/exec.Command` is passed non `*os.File` as an input/output, go
will create `os.Pipe`s and wait for their closure in `cmd.Wait()`.
If the code following this is responsible for closing `io.Pipe`s or
other handlers then on process death from context cancellation the
`Wait` can hang.

There are two possible solutions:

1. use `os.Pipe` as the input/output as `cmd.Wait` does not wait for these.
2. create a goroutine waiting on the context cancellation that will close the inputs.

This PR changes the inputs to the `CatFileBatch` calls to use `os.Pipe`s preventing
dangling cat-file calls.

Close #19454

Signed-off-by: Andrew Thornton <art27@cantab.net>

zeripath commented:
there are other places that need similar changes but this is the most important of these

Gusted commented on modules/git/batch_reader.go:
```diff
@@ -44,10 +35,28 @@ func EnsureValidGitRepository(ctx context.Context, repoPath string) error {
+		log.Critical("Unable to open pipe to write to: %v", err)
+		return returnClosedReaderWriters(err)
+	}
+	batchStdoutReader, batchStdoutWriter, err := os.Pipe()
```
We might want to check if this is CPU/Memory-efficient, the current implementation use buffered I/O(for good reason, this can and will be called on bigger files).

wxiaoguang replied:
`os.Pipe` is the correct and proper way. According to Golang source, if the `Stdout` provided is not a `os.File(Pipe)`, then there will be a new internal `os.Pipe` be used in the end.

So if you can provide the os.Pipe to Stdout, it will be used directly.


```go
func (c *Cmd) writerDescriptor(w io.Writer) (f *os.File, err error) {
	if w == nil {
		f, err = os.OpenFile(os.DevNull, os.O_WRONLY, 0)
		if err != nil {
			return
		}
		c.closeAfterStart = append(c.closeAfterStart, f)
		return
	}

	if f, ok := w.(*os.File); ok {
		return f, nil
	}

	pr, pw, err := os.Pipe()
	if err != nil {
		return
	}

	c.closeAfterStart = append(c.closeAfterStart, pw)
	c.closeAfterWait = append(c.closeAfterWait, pr)
	c.goroutine = append(c.goroutine, func() error {
		_, err := io.Copy(w, pr)
		pr.Close() // in case io.Copy stopped due to write error
		return err
	})
	return pw, nil
}
```

Gusted commented on modules/git/utils_closererror.go:
```diff
@@ -0,0 +1,99 @@
+	if c.Closer != nil {
+		_ = c.Closer.Close()
+	}
+	c.err = err
```
To be more in-line with current implementation:
```suggestion
	if err == nil {
		err = io.ErrClosedPipe
	}
	c.err = err	
```

wxiaoguang commented:
Then there is only one `nio` usage left, should it be refactored together? Then maybe the `nio` vendor package can be removed.

```go
// LogNameStatusRepo opens git log --raw in the provided repo and returns a stdin pipe, a stdout reader and cancel function
func LogNameStatusRepo(ctx context.Context, repository, head, treepath string, paths ...string) (*bufio.Reader, func()) {
	// We often want to feed the commits in order into cat-file --batch, followed by their trees and sub trees as necessary.
	// so let's create a batch stdin and stdout
	stdoutReader, stdoutWriter := nio.Pipe(buffer.New(32 * 1024))
```

wxiaoguang commented on modules/git/utils_closererror.go:
```diff
@@ -0,0 +1,102 @@
+	if c.Closer != nil {
+		_ = c.Closer.Close()
+	}
+	if err == nil {
+		err = io.ErrClosedPipe
+	}
```
It breaks the behavior of `os.PipeWriter`, the behavior is:

```
// CloseWithError closes the writer; subsequent reads from the
// read half of the pipe will return no bytes and the error err,
// or EOF if err is nil.
//
// CloseWithError never overwrites the previous error if it exists
// and always returns nil.
func (w *PipeWriter) CloseWithError(err error) error {
	return w.p.closeWrite(err)
}

func (p *pipe) closeWrite(err error) error {
	if err == nil {
		err = EOF
	}
	p.werr.Store(err)
	p.once.Do(func() { close(p.done) })
	return nil
}
```

zeripath replied:
Hmm... I'm just gonna revert this - we never looked for ErrClosedPipe or its string error type and I'm not overloading the Close() to call CloseWithError().

zeripath commented:
> Then there is only one `nio` usage left, should it be refactored together? Then maybe the `nio` vendor package can be removed.

Yes this was a planned follow-up, however:

There were (sometimes quite significant) performance improvements when the `nio.Pipe` was used over the `io.Pipe` implementation - (albeit the `io.Pipe` was itself a wrapper around an `os.Pipe`.) 

We're going to need to double check with some large files and large repositories to see whether a `nio.Pipe` wrapper around the `os.Pipe` is necessary. (The git or bsd ports repos are good test cases.)

There are a few options:

1. Just remove the `nio.Pipe` here - and in its backport - merge and see if there are apparent perf issues in real life usage. In a follow-up PR we could add back in the `nio.Pipe` if necessary. Meanwhile I continue working through the rest of the commands and remove the wrapping `io.Pipe`s and `nio.Pipe`s simplifying to `os.Pipe`
2. We attempt to get everything sorted in one PR.

Unfortunately, I don't know how Windows and BSD are going to respond to the loss of the buffering `nio.Pipe`. I suspect it will actually turn out to be an improvement but I'm not certain as it depends on the context switching of the kernel vs go's context switching.

There's another approach to solving the dangling file problem which makes fewer changes and you may prefer that to be merged and backported first.

wxiaoguang commented:
The `nio` was introduced in https://github.com/go-gitea/gitea/pull/16059 , but I think it's not necessary.

Pipe is a very general mechanism provided by OS, in most cases it could be treated as a socket (indeed they have the similar performance). Since we never add a buffer to a socket when copying streams, I don't think it's necessary to add a buffer to a pipe, either.

I prefer to drop the `nio` package.

zeripath commented on modules/git/repo.go:
```diff
@@ -329,18 +329,7 @@ func (repo *Repository) CreateBundle(ctx context.Context, commit string, out io.
-		return err
-	}
-	defer fi.Close()
+	_, _, err = NewCommand(ctx, "bundle", "create", "-", "bundle", "HEAD").RunStdString(&RunOpts{Dir: tmp, Env: env, Stdout: out})
```
I am uncertain why this code was written to use a temporary file in the first place. I suspect it may have been affected by this dangling problem and a temporary file was used to get around this?

zeripath commented:
OK, I've made substantial changes and created a common git.Pipe command with an implementation of PipeWriter and PipeReader that is based on the os.Pipe files.

If a git.PipeWriter or git.PipeReader (as appropriate) is passed in to a git.Command RunContext then the File() will be passed to the underlying os/exec process transparently. I'm not sure if this will/could have effects on error propagation but I think it will be fine. (Although I will check.)

I've removed the `nio.Pipe` stuff as we may not need it and we could always just add it back in - however, we'll only know about it's need once 1.17 is released because I do not think we can really backport this.

Most of the cases switching to using a git.Pipe won't have been affected by the dangling file case - certainly the pipeline functions I think were safe. However, it's still likely to be a performance improvement switching to use these. 

There are also a few cases where there was unnecessary usage of pipes and one case where I've dropped using a temporary file in preference to feeding the output down straight down the writer.

As an aside we need to come up with a succinct way of expressing when this kind of deadlock can occur and how to ensure it does not happen. Although we could simply change the Command RunContexts to only take git.PipeWriter and git.PipeReader as appropriate it might be better to express when exactly you need to have the additional: 

```go
go func() {
   <-ctx.Done()
   cancel()
   _ = stdin.Close()
   _ = stdout.Close()
}()
```

Perhaps actually Command should do this itself?

wxiaoguang commented:
Maybe we can introduce a `git.Pipes(n) ([]PipePair, err)` to save a lot of code, and the caller don't need to `Close` when error occurs

```go
pipes, err := Pipes(3)
if err != nil { ... } 
rd1, wr1 := pipes[0].rd, pipes[0].wr
rd2, wr2 := pipes[1].rd, pipes[1].wr
...
```

6543 commented:
as per https://github.com/go-gitea/gitea/pull/19454#pullrequestreview-950248088

wxiaoguang commented on modules/lfs/pointer_scanner_nogogit.go:
```diff
@@ -23,9 +23,28 @@ import (
-	catFileCheckReader, catFileCheckWriter := io.Pipe()
-	shasToBatchReader, shasToBatchWriter := io.Pipe()
-	catFileBatchReader, catFileBatchWriter := io.Pipe()
+	fail := func(err error) {
+		errChan <- err
+		close(pointerChan)
+		close(errChan)
+	}
+
+	pipes, err := git.NewPipePairs(3)
+	if err != nil {
+		fail(err)
+		return
+	}
+
+	closeAll := func(err error) {
+		for _, closer := range pipes {
+			_ = closer.CloseWithError(err)
+		}
+	}
+	defer closeAll(nil)
+
```
```suggestion
	var morePipes []*git.*PipePair
	var err error

	oldGit := git.CheckGitVersionAtLeast("2.6.0") != nil
	if oldGit {
		pipes, err = git.NewPipePairs(3+2)
	} else {
		pipes, err = git.NewPipePairs(3)
	}
	
	if err != nil {
		errChan <- err
		close(pointerChan)
		close(errChan)
		return
	}

	defer func() {
		for _, closer := range pipes {
			_ = closer.Close()
		}
	}()

```

wxiaoguang replied:
ps maybe there could also be a `git.ClosePipes(pipes, err)` function to close all pipes, I have seen the `for _, closer := range pipes { _ = closer.Close() }` a few times

wxiaoguang commented on modules/lfs/pointer_scanner_nogogit.go:
```diff
@@ -44,8 +63,20 @@ func SearchPointerBlobs(ctx context.Context, repo *git.Repository, pointerChan c
 
 	// 1. Run batch-check on all objects in the repository
 	if git.CheckGitVersionAtLeast("2.6.0") != nil {
```
```suggestion
	if oldGit {
```

wxiaoguang commented on modules/lfs/pointer_scanner_nogogit.go:
```diff
@@ -44,8 +63,20 @@ func SearchPointerBlobs(ctx context.Context, repo *git.Repository, pointerChan c
 	if git.CheckGitVersionAtLeast("2.6.0") != nil {
-		revListReader, revListWriter := io.Pipe()
-		shasToCheckReader, shasToCheckWriter := io.Pipe()
+
+		morePipes, err := git.NewPipePairs(2)
+		if err != nil {
+			wg.Done()
+			closeAll(err)
+			wg.Wait()
+			fail(err)
+			return
+		}
+
+		revListReader, revListWriter := morePipes[0].ReaderWriter()
+		shasToCheckReader, shasToCheckWriter := morePipes[1].ReaderWriter()
+		pipes = append(pipes, morePipes...)
```
```suggestion
		revListReader, revListWriter := morePipes[3].ReaderWriter()
		shasToCheckReader, shasToCheckWriter := morePipes[4].ReaderWriter()
```

wxiaoguang commented on modules/git/blob_nogogit.go:
```diff
@@ -54,11 +55,14 @@ func (b *Blob) DataAsync() (io.ReadCloser, error) {
 		cancel: cancel,
-	}, nil
+	}
+	runtime.SetFinalizer(br, func(br *blobReader) { br.Close() })
```
Hmm, these `SetFinalizer` have been extracted into a new PR 😊

lunny replied:
It will only be released when application exit. I don't think it's an expected behaviour?

zeripath replied:
No - the SetFinalizer will run when the object is gc'd:

> SetFinalizer sets the finalizer associated with obj to the provided finalizer function. When the garbage collector finds an unreachable block with an associated finalizer, it clears the association and runs finalizer(obj) in a separate goroutine. This makes obj reachable again, but now without an associated finalizer. Assuming that SetFinalizer is not called again, the next time the garbage collector sees that obj is unreachable, it will free obj.

wxiaoguang replied:
SetFinalizer shouldn't be used for production release. At least there should be an error log to remind developers that they forget to free resources.

zeripath replied:
Let's take this to the extracted PR. I've removed this from here.

Gusted replied:
The finalizer functionallity is not that uncommon to see with these kind of "long-lived" objects: https://github.com/golang/go/blob/d29f5247b8cbf5f2cb7b0e325a5eb1c7c5c1a91f/src/os/file_unix.go#L186

zeripath commented:
Looks like the race detector doesn't like the PipePair.Close() function because it has to close both ends of the pipes in different go-routines to the ones that read them. 

This is frustrating.

wxiaoguang commented:
Use `PipeReader.p.lock` to protect?

zeripath commented:
We'd need to lock the reads as well.
